include/hdf5libs/HDF5RawDataFile.hpp

```diff
@@ -0,0 +1,211 @@
+/**
+ * @file HDF5RawDataFile.hpp
+ *
+ * Class for reading out HDF5 files as
+ *
+ *
+ * This is part of the DUNE DAQ Application Framework, copyright 2020.
+ * Licensing/copyright details are in the COPYING file that you should have
+ * received with this code.
+ */
+
+#ifndef HDF5LIBS_INCLUDE_HDF5LIBS_HDF5RAWDATAFILE_HPP_
+#define HDF5LIBS_INCLUDE_HDF5LIBS_HDF5RAWDATAFILE_HPP_
+
+// DUNE-DAQ
+#include "hdf5libs/hdf5filelayout/Structs.hpp"
+#include "hdf5libs/HDF5FileLayout.hpp"
+
+#include "daqdataformats/Fragment.hpp"
+#include "daqdataformats/TriggerRecord.hpp"
+#include "logging/Logging.hpp"
+
+// External Packages
+#include <highfive/H5DataSet.hpp>
+#include <highfive/H5File.hpp>
+#include <highfive/H5Group.hpp>
+#include <highfive/H5Object.hpp>
+#include <nlohmann/json.hpp>
+
+// System
+#include <filesystem>
+#include <iostream>
+#include <map>
+#include <memory>
+#include <set>
+#include <string>
+#include <sys/statvfs.h>
+#include <variant>
+#include <vector>
+
+namespace dunedaq {
+
+ERS_DECLARE_ISSUE(hdf5libs,
+                  FileOpenFailed,
+                  "Issue when opening file " << file << ": " << message,
+                  ((std::string)file)((std::string)message))
+
+namespace hdf5libs {
+
+/**
+ * @brief HDF5RawDataFile is the class responsible
+ * for interfacing the DAQ format with the HDF5 file format.
+ */
+class HDF5RawDataFile
+{
+
+public:
+  enum
+  {
+    TLVL_BASIC = 2,
+    TLVL_FILE_SIZE = 5
+  };
+
+  // constructor for writing
+  HDF5RawDataFile(std::string file_name,
+                  daqdataformats::run_number_t run_number,
+                  size_t file_index,
+                  std::string application_name,
+                  const hdf5filelayout::FileLayoutParams& fl_params,
+                  unsigned open_flags = HighFive::File::Create);
+  HDF5RawDataFile(std::string file_name,
+                  daqdataformats::run_number_t run_number,
+                  size_t file_index,
+                  std::string application_name,
+                  const nlohmann::json& fl_params_conf,
+                  unsigned open_flags = HighFive::File::Create);
+
+  // constructor for reading
+  explicit HDF5RawDataFile(const std::string& file_name);
+
+  ~HDF5RawDataFile();
+
+  std::string get_file_name() const { return m_file_ptr->getName(); }
+
+  size_t get_recorded_size() const noexcept { return m_recorded_size; }
+
+  HDF5FileLayout get_file_layout() const { return *(m_file_layout_ptr.get()); }
+
+  // basic data writing methods
+  void write(const daqdataformats::TriggerRecord& tr);
+  void write(const daqdataformats::TriggerRecordHeader& trh);
+  void write(const daqdataformats::Fragment& frag);
+
+  // attribute writers/getters
+  template<typename T>
+  void write_attribute(std::string name, T value)
```

Comment:
Move definition out of the class declaration,  to the bottom of the header file.